File: ACC22.py
import numpy as np
import networkx as nx
from scipy.optimize import minimize

class Agent:
    def __init__(self, initial_value, convex_set):
        self.value = initial_value 
        self.convex_set = convex_set 
    
    def filter_values(self, received_values, f):
        distances = [np.linalg.norm(self.value - rv) for rv in received_values]
        filtered_indices = np.argsort(distances)[:-f]
        return [received_values[i] for i in filtered_indices]
    
    def project_to_convex_set(self, x):
        A, b = self.convex_set
        cons = [{'type': 'ineq', 'fun': lambda x, A=A[i], b=b[i]: b - np.dot(A, x)} for i in range(len(b))]
        result = minimize(lambda x: np.linalg.norm(x - self.value), x, constraints=cons)
        return result.x
    
    def update_value(self, retained_values, alpha):
        update_step = np.sum([rv - self.value for rv in retained_values], axis=0)
        self.value += alpha * update_step
        self.value = self.project_to_convex_set(self.value)  

class ConsensusSystem:
    def __init__(self, agents, alpha, f):
        self.agents = agents
        self.alpha = alpha
        self.f = f
    
    def run_consensus(self, num_iterations):
        for _ in range(num_iterations):
            new_values = []
            for agent in self.agents:
                received_values = [a.value for a in self.agents if a != agent]
        
                retained_values = agent.filter_values(received_values, self.f)
               
                agent.update_value(retained_values, self.alpha)
            
            new_values.append([agent.value for agent in self.agents])
        
        return new_values

agents = [
    Agent(np.random.rand(2), (np.eye(2), np.ones(2))) for _ in range(5)
]  
consensus_system = ConsensusSystem(agents, alpha=0.1, f=1)

consensus_values = consensus_system.run_consensus(10)

for i, agent in enumerate(agents):
    print(f"Agent {i+1} final value: {agent.value}")

File: ACC22.m
classdef Agent
    properties
        value
        convex_set
    end
    
    methods
        function obj = Agent(initial_value, convex_set)
            obj.value = initial_value;
            obj.convex_set = convex_set;
        end
        
        function retained_values = filter_values(obj, received_values, f)
            distances = vecnorm(obj.value - received_values, 2, 2);
            [~, sorted_indices] = sort(distances);
            retained_values = received_values(sorted_indices(1:end-f), :);
        end
        
        function projected_value = project_to_convex_set(obj, x)
            A = obj.convex_set.A;
            b = obj.convex_set.b;
            options = optimoptions('fmincon', 'Display', 'off');
            projected_value = fmincon(@(y) norm(y - x), x, A, b, [], [], [], [], [], options);
        end
        
        function obj = update_value(obj, retained_values, alpha)
            update_step = sum(retained_values - obj.value, 1);
            obj.value = obj.value + alpha * update_step;
            obj.value = obj.project_to_convex_set(obj.value);
        end
    end
end

classdef ConsensusSystem
    properties
        agents
        alpha
        f
    end
    
    methods
        function obj = ConsensusSystem(agents, alpha, f)
            obj.agents = agents;
            obj.alpha = alpha;
            obj.f = f;
        end
        
        function new_values = run_consensus(obj, num_iterations)
            for iter = 1:num_iterations
                for i = 1:length(obj.agents)
                    agent = obj.agents(i);
                    received_values = vertcat(obj.agents([1:i-1, i+1:end]).value);
                    retained_values = agent.filter_values(received_values, obj.f);
                    obj.agents(i) = agent.update_value(retained_values, obj.alpha);
                end
            end
            
            new_values = vertcat(obj.agents.value);
        end
    end
end
